underfs/cos/src/main/java/alluxio/underfs/cos/COSOutputStream.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs.cos;
+
+import alluxio.util.io.PathUtils;
+
+import com.google.common.base.Preconditions;
+import org.apache.commons.codec.binary.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.qcloud.cos.COSClient;
+import com.qcloud.cos.model.ObjectMetadata;
+import com.qcloud.cos.exception.CosClientException;
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A stream for writing a file into COS. The data will be persisted to a temporary directory on the
+ * local disk and copied as a complete file when the {@link #close()} method is called.
+ */
+@NotThreadSafe
+public final class COSOutputStream extends OutputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(COSOutputStream.class);
+
+  /** Bucket name of the Alluxio COS bucket. */
+  private final String mBucketName;
+  /** Key of the file when it is uploaded to COS. */
+  private final String mKey;
+  /** The local file that will be uploaded when the stream is closed. */
+  private final File mFile;
+  /** The cos client for COS operations. */
+  private final COSClient mCosClient;
+
+  /** The outputstream to a local file where the file will be buffered until closed. */
+  private OutputStream mLocalOutputStream;
+  /** The MD5 hash of the file. */
+  private MessageDigest mHash;
+
+  /** Flag to indicate this stream has been closed, to ensure close is only done once. */
+  private AtomicBoolean mClosed = new AtomicBoolean(false);
+
+  /**
+   * Creates a name instance of {@link COSOutputStream}.
+   *
+   * @param bucketName the name of the bucket
+   * @param key the key of the file
+   * @param client the client for COS
+   */
+  public COSOutputStream(String bucketName, String key, COSClient client) throws IOException {
+    Preconditions.checkArgument(bucketName != null && !bucketName.isEmpty(),
+        "Bucket name must not be null or empty.");
+    Preconditions.checkArgument(key != null && !key.isEmpty(),
+        "COS path must not be null or empty.");
+    Preconditions.checkArgument(client != null, "COSClient must not be null.");
+    mBucketName = bucketName;
+    mKey = key;
+    mCosClient = client;
+
+    mFile = new File(PathUtils.concatPath("/tmp", UUID.randomUUID()));
```

Comment:
`mFile = new File(PathUtils.concatPath(CommonUtils.getTmpDir(), UUID.randomUUID()));`, in case the tmp dir is not `/tmp`